--- portage/bin/ebuild-helpers/prepstrip	2011-03-10 16:24:48.000000000 +0100
+++ prepstrip	2011-03-10 16:34:33.000000000 +0100
@@ -149,6 +149,11 @@
 	# actually causes problems.  install sources for all
 	# elf types though cause that stuff is good.
 
+	if type -P getcap > /dev/null ; then
+		#save caps
+		getcapOutput=`getcap "${x}"` 
+	fi
+
 	if [[ ${f} == *"current ar archive"* ]] ; then
 		vecho "   ${x:${#D}}"
 		if ${strip_this} ; then
@@ -170,6 +175,14 @@
 			${STRIP} ${SAFE_STRIP_FLAGS} "${x}"
 		fi
 	fi
+
+	if type -P setcap > /dev/null ; then
+		#Apply saved caps
+		capset=${getcapOutput#*= }
+		if [ -n "$capset" ] ; then
+			setcap $capset "${x}" 
+		fi
+	fi
 done
 
 if [[ -s ${T}/debug.sources ]] && \
